### PR TITLE
Frontend refactor

### PR DESCRIPTION
**Pull request by allenwq**
- Implement a way to do page specific javascript, the javascript will only be required on the specific page. Thus there's no need to use `DOCUMENT_SELECTOR` to scope the events anymore.
- Split out `Intl` polyfill, now `Intl` will only be downloaded if the browser does not support it ( This reduces the JS bundle size by 150K ).
- Remove turbolinks
  Turbolinks does make the page load faster, but it has some downsides:
  - The global scope does not get cleaned since JS are not reloaded, one page will have all listeners from all JS in the app. This will finally make the frontend slow if we have more and more JS. And it also requires we take care of every piece of JS and make sure they are scoped to a specific page.
  - It does not work with frontend code split and async libraries.
  - It will be likely replaced with react-router later since we are moving more and more parts to react.
  - [See more](https://shakacode.gitbooks.io/react-on-rails/content/docs/additional-reading/turbolinks.html#requirements-for-using-turbolinks)
- Remove react_on_rails
  - Discussed with @kxmbrian in #1608
  -  React on rails assumes a single big bundle and also does not support [conditional load javascript libraries](https://webpack.github.io/docs/code-splitting.html). 
  - It couples the frontend with rails, which makes swapping of the backend hard.

### CODE REVIEW

**Comment by allenwq** [comment]
I am going to port the existing javascript to webpack side and use ES6 next. Think need to restructure the folders a bit. 

There's a `helpers` folder on rails side, I might move it to `client/app/helpers` or `client/app/lib`, @kxmbrian @weiqingtoh any thoughts ?

**Comment by allenwq** [comment]
Have rewritten the table row form helpers using ES6 in the latest commit as an example, reduced a lot of duplication!

File: client/scripts/startup.js [hunk 1]
```diff
@@ -1,13 +1,24 @@
...
-  start();
+  require.ensure([], (require) => {
+    require('intl');
+    require('intl/locale-data/jsonp/en');
+    require('intl/locale-data/jsonp/zh');
+    loadCurrentModule();
+  }, 'intl');
 } else {
-  start();
+  loadCurrentModule();
```
Comment (weiqingtoh) [inline client/scripts/startup.js]: Why not bring `loadCurrentModule()` out of both blocks? 

Will we have cases of other packages like `intl`, where the dependency cannot be inferred by the `server_meta_tag`?

Comment (allenwq) [inline client/scripts/startup.js]: The two has different contexts. The first one is in the async callback, only there you are sure the Intl is required correctly. 

Yah, it's possible if we have such polyfills in future, now only Safari <= 9 does not support Intl.

File: client/webpack.config.js [hunk 2]
```diff
@@ -12,7 +12,9 @@ const config = {
 
   output: {
     filename: 'coursemology.bundle.js',
+    chunkFilename: '[name]-[chunkhash].bundle.js',
```
Comment (weiqingtoh) [inline client/webpack.config.js]: Just to check, where is the `[chunkhash]` defined?

File: client/scripts/startup.js [hunk 3]
```diff
@@ -1,13 +1,24 @@
...
-function start() {
-  require('../app/bundles/Course/LessonPlan/startup');
+/* eslint-disable import/no-dynamic-require */
+
+function loadCurrentModule() {
+  const modulePath = document.head.querySelector("[name='server-context']")
+                       .getAttribute('data-controller-name');
+
+  try {
+    require(`../app/bundles/${modulePath}`);
```
Comment (weiqingtoh) [inline client/scripts/startup.js]: From this, if there are common javascript functions, they should be placed in `/helpers`, and imported if required?

**Comment by weiqingtoh** [comment]
👍 The refactor seems like a good move in the right direction - turbolinks + single asset pipeline might not be good for a large app like Coursemology. 

Once we update this, it might be good to update either `CONTRIBUTING.md` or `/client/README.md` (or both) on the conventions.

File: client/scripts/startup.js (hunk 1 above)
Comment (weiqingtoh) [inline client/scripts/startup.js]: 👍

File: client/webpack.config.js (hunk 2 above)
Comment (allenwq) [inline client/webpack.config.js]: It's like a special placeholder from webpack: https://webpack.github.io/docs/long-term-caching.html

they will be replaced with the hash of the file after compiling.

File: client/scripts/startup.js (hunk 3 above)
Comment (allenwq) [inline client/scripts/startup.js]: I am following our existing convention and put exiting `helpers` to `client/lib/helpers`, maybe the folders can be further renamed or better structured.

File: client/app/bundles/course/lesson_plan/items.jsx [hunk 4]
```diff
@@ -1,10 +1,38 @@
...
+import { IntlProvider, addLocaleData } from 'react-intl';
+
+import zh from 'react-intl/locale-data/zh';
+
+import createStore from './store';
+import LessonPlanContainer from './containers/LessonPlanContainer';
+import translations from '../../../../build/locales/locales.json';
+
+function renderLessonPlan(props) {
+  const i18nLocale = $("meta[name='server-context']").data('i18n-locale');
```
Comment (weiqingtoh) [inline client/app/bundles/course/lesson_plan/items.jsx]: Not sure if it's possible, but can we handle locales in a helper?

**Comment by allenwq** [comment]
The latest commits further restructure the files and renamed `startup.js` to `app/index.js`. The things inside scripts are the task like scripts which are different from app entry file.

Thanks @weiqingtoh for helping with the `CONTRIBUTION.md` !

File: client/app/bundles/course/lesson_plan/items.jsx (hunk 4 above)
Comment (allenwq) [inline client/app/bundles/course/lesson_plan/items.jsx]: Yep, I think that's way to go. We should factor out the Intl logic to a separate container. I will defer this when implementing the next component.

**Pull request merged**